src/Common/src/System/Net/_LazyAsyncResult.cs

```diff
@@ -0,0 +1,523 @@
+                            (result is Exception ? ((Exception)result).Message : result == null ? "<null>" : result.ToString()) +
+                            ", userToken:" + userToken.ToString());
+
+            // Critical to disallow DBNull here - it could result in a stuck spinlock in WaitForCompletion.
+            if (result == DBNull.Value)
+            {
+                throw new ArgumentNullException("result");
+            }
+
+#if DEBUG
+            // Always safe to ask for the state now.
+            _ProtectState = false;
+#endif
+
+            if ((_intCompleted & ~c_HighBit) == 0 && (Interlocked.Increment(ref _intCompleted) & ~c_HighBit) == 1)
+            {
+                // DBNull.Value is used to guarantee that the first caller wins,
+                // even if the result was set to null.
+                if (_result == DBNull.Value)
+                    _result = result;
+
+                ManualResetEvent asyncEvent = (ManualResetEvent)_event;
+                if (asyncEvent != null)
+                {
+                    try
+                    {
+                        asyncEvent.Set();
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                        // Simply ignore this exception - There is apparently a rare race condition
+                        // where the event is disposed before the completion method is called.
+                    }
+                }
+
+                Complete(userToken);
+            }
+        }
+
+        // A method for completing the IO with a result and invoking the user's callback.
+        internal void InvokeCallback(object result)
+        {
+            ProtectedInvokeCallback(result, IntPtr.Zero);
+        }
+
+        // A method for completing the IO without a result and invoking the user's callback.
+        internal void InvokeCallback()
+        {
+            ProtectedInvokeCallback(null, IntPtr.Zero);
+        }
+
+        //
+        //  MUST NOT BE CALLED DIRECTLY
+        //  A protected method that does callback job and it is guaranteed to be called exactly once.
+        //  A derived overriding method must call the base class somewhere or the completion is lost.
+        //
+        protected virtual void Complete(IntPtr userToken)
+        {
+            bool offloaded = false;
+            ThreadContext threadContext = CurrentThreadContext;
+            try
+            {
+                ++threadContext.m_NestedIOCount;
+                if (_asyncCallback != null)
+                {
+                    GlobalLog.Print("LazyAsyncResult#" + Logging.HashString(this) + "::Complete() invoking callback");
+
+                    if (threadContext.m_NestedIOCount >= c_ForceAsyncCount)
+                    {
+                        GlobalLog.Print("LazyAsyncResult::Complete *** OFFLOADED the user callback ***");
+                        Task.Factory.StartNew(WorkerThreadComplete, null);
```

Comment:
Per previous comment, this would be better as:
```C#
Task.Factory.StartNew(s => ((LazyAsyncResult)s).WorkerThreadComplete(), this,
    CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```